One Way Data flow & Two Scrollbars

alexander-alvarez opened the pull request:
I gutted the `ember-scrollable` class in favor of a more DDAU approach. Allowing me to make the double scroll easily.

![image](https://cloud.githubusercontent.com/assets/11779993/21412237/bcac4396-c7bc-11e6-97f3-7b633328f689.png)


This made it easier to set up listeners on the appropriate 'information-hierarchy' level and delegate functionality to different components

new components (ideally would be private):
`scroll-content-element` -- Handles scroll events within the body of the scroll content, properly sets scrollTop / scrollLeft property depending on the configuration and the given `scrollTo`.
`ember-scrollbar` --  Handles displaying and moving the handle within the confines of it's template, and has callbacks for intending to dragging and jump to particular positions.

**TODO**

- [x] ~Extend api to allow for left and right scroll & implement~
- [x] ~Fix Jiterryness~
- [x] ~Lots of documentation~
- [x] ~Tests~

**Considerations**
- I couldn't figure out how to completely eliminate the ember-scrollable class, so there is parent's accessing children's state through jquery.
```hbs
<!-- singleScroll -->
{{ember-scrollable
  horizontal=true
  scrollTo=0
}}

<!-- double scroll -->
{{ember-scrollable
  horizontal=true
  vertical=true
  scrollToX=0
  scrollToY=0
}}

```


End goal is something like this (except for ours hides when the user isn't moving the mouse in the scroll area):
http://automattic.github.io/antiscroll/

taras commented:
I just wanted to let you know that I started looking at the code and it's looking great so far. I haven't had a chance to pull it down yet and take a deeper look.

alexander-alvarez commented:
Thanks @taras . Just got the two scrollbars to show up and trigger appropriate scroll events.
now to debug jittery-ness and iron out the kinks.

>  `
{{ember-scrollable
  horizontal=true
  scrollTo=0
}}

> `
{{ember-scrollable
  double=true
  scrollToX=0
  scrollToY=0
}}

I decided I liked the first approach. I'm thinking `scrollTo` would map  to the appropriate `scrollToX` or `scrollToY` depending on what the value for `horizontal` is set to, that way there's no need to change any of the implementation on the `scroll-content-element`.

alexander-alvarez commented:
out of curiousity -- you wouldn't happen to know if this file is needed: https://github.com/alphasights/ember-scrollable/blob/master/blueprints/ember-scrollable/index.js

it's the only piece of code I see that references trackpad scroll emulator

taras commented:
I'm not sure if it's necessary because it's now added as a dependency. I
don't remember, it's been a while. My only concern would be whether it
would work in ember-light-table without that code.

-- 
Taras Mankovski

On December 9, 2016 at 9:42:50 AM, Alex Alvarez (notifications@github.com)
wrote:

> out of curiousity -- you wouldn't happen to know if this file is needed:
> https://github.com/alphasights/ember-scrollable/blob/master/blueprints/ember-scrollable/index.js
>
> it's the only piece of code I see that references trackpad scroll emulator
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/alphasights/ember-scrollable/pull/29#issuecomment-265929350>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEjvwUUdXR8YTdcrwkwbUTLJEJmGxsEks5rGNVCgaJpZM4LG6nD>
> .
>

alexander-alvarez commented:
**Planned tests**
- [ ] Using mousewheel triggers vertical scroll
- [ ] Click on the scrollbar when not on handle causes 'jumping' to particular position
- [ ] Click and drag on scrollbar space moves the scrollbar (in vertical and horizontal mode). After mouseUp (anywhere) dragging should stop
- [ ] Click, move mouse off of scrollbar, and drag moves the scrollbar. After mouseup, dragging should stop.

alexander-alvarez commented on tests/integration/components/scroll-content-element-test.js:
```diff
@@ -0,0 +1,104 @@
+  const secondPosition = 25;
+
+  // WHEN the scrollY position has moved 10 px
+  this.$(cssSelector).scrollTop(firstPosition);
```
@taras I was struggling to get these badboys triggering the scroll events synchronously. 
the source binding the scroll is as such: `this.$().on('scroll', bind(this, this.scrolled));`

I tried wrapping these blocks in `run`, `run.next` etc, to no avail.

any experience with this?

alexander-alvarez commented:
The code is ready for UAT. @offirgolan  @taras @buschtoens 

Docs and tests, and fixing any bugs you may find are all that's left.

offirgolan commented:
@alexander-alvarez this looks great, the only thing I'm not very keen on is the naming for the `double` scrollbar. I think it would be more intuitive to have something like

```hbs
{{ember-scrollable
  horizontal=true
  vertical=true
  scrollTo=0
}}
```

This way you can also bind properties to toggle horizontal and/or vertical scroll.

offirgolan commented:
Also, I'm not too sure how this would work with ember light table since we can't just enable vertical AND horizontal scroll on the `tbody`. It would need to be something like this:

```hbs
{{#ember-scrollable vertical=true}}
  <table>
    <thead></thead>
  <table>

  {{#ember-scrollable horizontal=true}}
    <table>
      <tbody><tbody>
    </table>
  {{/ember-scrollable}}

{{/ember-scrollable}}

alexander-alvarez commented:
@offirgolan 
I agree with the api decision to not use `double`

What about with `scrollTo`? 
```hbs

{{!-- not valid? default to vertical? --}}
{{#ember-scrollable
  horizontal=true
  vertical=true
  scrollTo=0
}}

{{!-- detect horizontal is set and proxy to scrollToX--}}
{{#ember-scrollable
  horizontal=true
  scrollTo=0
}}

{{!-- detect vertical is set and proxy to scrollToY --}}
{{#ember-scrollable
  vertical=true
  scrollTo=0 
}}

{{!-- explicit form --}}
{{#ember-scrollable
  horizontal=true
  vertical=true
  scrollToX=0 
  scrollToY=0 
}}


```

offirgolan commented:
@alexander-alvarez I like the idea of replacing `scrollTo` with `scrollToX` and `scrollToY`. I think its much more fitting and more explicit. Would it be a difficult task to separate the two? Maybe @buschtoens can help out on that as well 😄

alexander-alvarez commented:
> I like the idea of replacing scrollTo with scrollToX and scrollToY. I think its much more fitting and more explicit. Would it be a difficult task to separate the two?

@offirgolan  it's already done, as is the proxying of `scrollTo` to `scrollToX` or `scrollToY` depending on what direction the scrollbar is going in.
see https://github.com/alphasights/ember-scrollable/pull/29/files#diff-499447ac5ce8a7c1077b7503f9403d0dR65


We'll have to see how to integrate the scrollbars with light-table.. also excited about fixed columns and rows and how it all interacts

taras commented on tests/integration/components/ember-scrollbar-test.js:
```diff
@@ -0,0 +1,385 @@
+    </div>
+  </div>`);
+
+  // TODO no idea why offset and width are showing up as half of the set properties
```
This is happening because app container is set to zoom: 50%

alexander-alvarez replied:
where is the ungodly piece of code that does this so I can reference some sort of variable or something istead of `/ 2`?
I was questioning my sanity during testing this :joy:

taras commented:
This is looking awesome! 👍

offirgolan commented:
@alexander-alvarez I think we should get rid of `scrollTo` altogether. I think it will just be confusing and I really dont like the idea of having to proxy the value. 

Also, I believe you should be able to set `scrollToX` and `scrollToY` independently of each other no? Wouldnt that make more sense?

alexander-alvarez commented:
@offirgolan currently now you don't have to proxy it -- you could use the components how you've shown them above, and it will work -- I just added the proxying to keep backwards compatability... happy and easy to remove it, if we want to break compatability

offirgolan commented:
@alexander-alvarez that's fine. I would also add a deprecation in the `set` of the `scrollTo` property then. I'd prefer users to not directly set `scrollTo` and eventually remove the proxy altogether.

offirgolan commented:
@alexander-alvarez is this ready to be merged?

alexander-alvarez commented:
@offirgolan almost... 

- [ ] I was going to add the deprecation (which is simple)
- [ ] Tests are broken in master https://github.com/alphasights/ember-scrollable/pull/33
- [ ] Adjust for the zoom programatically (@taras need to understand how I get access to that programatically)
- [ ] I need help handling the asynchroneity of the scroll events... see the `run.later` in tests.

taras commented:
>  Adjust for the zoom programatically (@taras need to understand how I get access to that programatically)

Why do you want to adjust zoom programmatically?

> I need help handling the asynchroneity of the scroll events... see the run.later in tests.

We could use [`ember-lifeline`](https://github.com/rwjblue/ember-lifeline) to address this issue and #30 in a more general way. But this will make `ember-scrollable` require Ember 2.6+. Thoughts?

alexander-alvarez commented:
@taras  adjust *for* the zoom programatically. I want a variable instead of [`/ 2`](https://github.com/alphasights/ember-scrollable/pull/29/files#diff-17ee528b7a6ca07afaa4702fb54b6f7fR37)

right now we're doing [bind](https://github.com/alphasights/ember-scrollable/pull/29/files#diff-bcc853db8ecac305500add8bb7a0f313R170) , which I guess I can replace the with the code from https://github.com/rwjblue/ember-lifeline#addeventlistener if that's the best practice and will let me replace the arbitrary run.later's with run loops.

I don't think requiring 2.6+ is a little excessive at the moment, no?

taras commented:
> @taras adjust for the zoom programatically. I want a variable instead of / 2

I'm not sure how you'd handle that. We could remove zoom from the test suite. What do you think?

> I don't think requiring 2.6+ is a little excessive at the moment, no?

I'm not sure. I've seen `ember-lifeline` being used in other addons. I'm not sure how many people are using < Ember 1.13. Those people can just stick to an older version of Ember Scrollable

alexander-alvarez commented:
> We could remove zoom from the test suite

Any particular reason why it's there? Where is this even done?


oh wait.. I already added `ember-lifeline` :joy: 
https://github.com/alphasights/ember-scrollable/pull/29/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R39

taras commented:
> Any particular reason why it's there? Where is this even done?

It's part of the default app blueprint. It's in [`ember-qunit`](https://github.com/ember-cli/ember-cli-qunit/commit/2e2d0632dcf9ff94b5e7b9788a2e5efef19e61c5#diff-2357a87860b59b92e5897426c4905e10L14) but it looks like it was removed. We can upgrade to the latest `ember-qunit` and it'll just resolve itself (i think)

alexander-alvarez commented:
@taras  no luck
https://github.com/alphasights/ember-scrollable/pull/33/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R30

offirgolan commented:
@alexander-alvarez can't we just force the zoom to be `1` via css on `tests/index.html`?

```html
<style>
#ember-testing-contaner {
 zoom: 1 !important;
}
</style>
```

Note: I'm not sure if the id of the container is correct, just going off the top of my head here.

alexander-alvarez commented:
tried write a [test helper](https://github.com/alexander-alvarez/ember-scrollable/compare/one-way-flow-2...alexander-alvarez:one-way-with-test-helper), that blocks for the scroll events which are asynchronous, but still not passing -- I'm pretty sure it's related to the scroll events triggered by `configureInitialScrollPosition` and them not getting `flushed`

offirgolan commented:
@alexander-alvarez are these test being run on chrome or phantom?

alexander-alvarez commented:
@offirgolan travis it seems they're using chrome (on my local it's with chrome)

alexander-alvarez commented:
What needs to be done to finalize this? 
Do we or anyone we know how to properly block for the asynchronous scroll events in tests?

Test is failing because of:
https://github.com/alphasights/ember-scrollable/pull/33

alexander-alvarez commented:
I updated to follow the LTS releases supported by `ember-lifeline`: https://github.com/rwjblue/ember-lifeline/blob/master/.travis.yml#L15
added badge to README indicating 2.3+ support, got rid of the `ember-hash-helper-polyfill` stuff for ember 1.13 support.

offirgolan commented:
Looks like test are passing now as well. Awesome work @alexander-alvarez.

cc @taras. Lets get this baby in so we can start figuring out how to integrate this with ELT.

taras commented:
Amazing job here guys!!!
